<a href="https://colab.research.google.com/github/dm-fedorov/advanced-python/blob/master/floating_point_numbers.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open and Execute in Google Colaboratory" target="_blank"></a>

см. https://realpython.com/python-numbers/

### Числа в Python

Числа в языке Python представлены тремя встроенными типами: целые (`int`), вещественные (`float`) и комплексные (`comlex`), а так же двумя типами чисел, которые предоставляет его стандартная библиотека: десятичные дроби неограниченной точности (`Decimal`) и обыкновенные дроби (`Float`).

Начнем с того, что числовые литералы не содержат знаки "`+`" и "`-`", т.е. с точки зрения Python число `−3.14` не является единым отрицательным числом, а является командой, которая состоит из унарного оператора (`-`) и числа `3.14`. Это говорит о том, что знаки "`+`" и "`-`" хоть и способны обозначать положительные и отрицательные числа, но на самом деле эти знаки являются операторами:

In [1]:
-+-+--++1

1

Помимо этого, числа встроенных типов (`int`, `float` и `complex`) относятся к неизменяемым объектам, т.е. объектам, чья структура не может быть изменена напрямую. Нельзя изменить какую-нибудь цифру существующего числа, нельзя расставить его цифры в обратном порядке. То, что кажется изменением числа, на самом деле таковым не является. Например:

In [2]:
x = 1
x = x - 1
x

0

Вроде бы мы создали объект-число со значением `1`, а затем изменили его значение на `0`. Но если это так, то `id` объекта (его адрес в памяти) не должен меняться, а он меняется:

In [3]:
x = 1
id(x)

93956286143808

In [4]:
x = x - 1
x

0

In [5]:
id(x)

93956286143776

Как видим, изменения объекта не произошло, старый объект исчез и появился новый. 

Результат для встроенных числовых типов всегда преобразуется к более общему типу, если это необходимо. Например, если частное двух целых чисел (`int`) не является целым числом, то оно будет преобразовано к вещественному типу (`float`):

In [6]:
1 / 25

0.04

А если результат не может быть выражен типами `int` и `float`, то он будет преобразован к типу `complex`. Так что если вдруг будете вычислять корни четной степени из отрицательных чисел, не ждите ошибки:

In [7]:
(-3) ** 0.5

(1.0605752387249068e-16+1.7320508075688772j)

In [8]:
(-3.14) ** 0.5

(1.0850398284807605e-16+1.772004514666935j)

### Целые числа (int)

В самом общем смысле, целые числа - это самые обыкновенные целые числа со знаком или без, например: 
$−11$, $126$, $0$ или $401734511064747568885490523085290650630550748445698208825344$. 
Последнее число в примере - это $2^{198}$, в чем очень легко убедиться:

In [9]:
2 ** 198

401734511064747568885490523085290650630550748445698208825344

Например, вы можете легко убедиться в том что 561 - [число Кармайкла](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%BE_%D0%9A%D0%B0%D1%80%D0%BC%D0%B0%D0%B9%D0%BA%D0%BB%D0%B0), действительно, проходит [тест Ферма](https://ru.wikipedia.org/wiki/%D0%A2%D0%B5%D1%81%D1%82_%D0%A4%D0%B5%D1%80%D0%BC%D0%B0):

In [10]:
2 ** (561 - 1) % 561 == 1

True

Однако, если вы попытаетесь проверить это для числа $9746347772161$, 
то результата придется ждать очень долго (если вообще дождемся). 

Но, если воспользоваться встроенной функцией `pow()`, то результат будет получен моментально:

In [11]:
pow(2, 9746347772160, 9746347772161) == 1

True

Все дело в том, что данная функция для трех аргументов реализована, как [алгоритм быстрого возведения в степень по модулю](https://ru.wikipedia.org/wiki/%D0%90%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC%D1%8B_%D0%B1%D1%8B%D1%81%D1%82%D1%80%D0%BE%D0%B3%D0%BE_%D0%B2%D0%BE%D0%B7%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D1%8F_%D0%B2_%D1%81%D1%82%D0%B5%D0%BF%D0%B5%D0%BD%D1%8C_%D0%BF%D0%BE_%D0%BC%D0%BE%D0%B4%D1%83%D0%BB%D1%8E), что на порядки быстрее, чем эквивалентная команда:
```
2 ** 9746347772160 % 9746347772161
```

### Вещественные числа (float)

Наверное, было бы правильнее называть эти числа [числами с плавающей точкой](https://ru.wikipedia.org/wiki/%D0%A7%D0%B8%D1%81%D0%BB%D0%BE_%D1%81_%D0%BF%D0%BB%D0%B0%D0%B2%D0%B0%D1%8E%D1%89%D0%B5%D0%B9_%D0%B7%D0%B0%D0%BF%D1%8F%D1%82%D0%BE%D0%B9) нежели [вещественными](https://ru.wikipedia.org/wiki/%D0%92%D0%B5%D1%89%D0%B5%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D0%BE%D0%B5_%D1%87%D0%B8%D1%81%D0%BB%D0%BE), но в принципе, с определенной натяжкой, можно сказать, что это, как бы одно и тоже. Давайте разберемся почему.

В качестве примера возьмем число $\sqrt2$, которое является *вещественным*, потому что мы никогда не сможем выразить его с помощью обыкновенной дроби. А если мы все-таки извлечем корень из двойки, то обнаружим, что это бесконечная десятичная дробь. 

Но вычислив этот корень на Python:

In [12]:
2 ** (1 / 2)    #  равносильно 2**0.5

1.4142135623730951

Мы увидим, что никакой бесконечной дробью и не пахнет. Python вернул только начало этой дроби ([детали реализации на Python](https://docs.python.org/3/tutorial/floatingpoint.html)), а все остальное отбросил, т.е. он вернул число с плавающей точкой, которое как бы и соответствует вещественному числу, но с определенной погрешностью.

На самом деле, работая с числами с плавающей точкой, мы очень часто подразумеваем числа вещественные, например вот такое число $\sqrt[77]7$, его мы увидим в виде конечной десятичной дроби:

In [13]:
7 ** (1 / 77)

1.0255935932948266

А число $7^{-77}$ $-$ в виде [мантисы](https://neerc.ifmo.ru/wiki/index.php?title=%D0%9F%D1%80%D0%B5%D0%B4%D1%81%D1%82%D0%B0%D0%B2%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B2%D0%B5%D1%89%D0%B5%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D1%8B%D1%85_%D1%87%D0%B8%D1%81%D0%B5%D0%BB) $8.461569363277291$ (опять же конечной десятичной дроби) и порядка $−66$:

In [14]:
7 ** (-77)

8.461569363277291e-66

или так: $8.461569363277291*10^{-66}$

Кстати, можно было бы подумать, что `8.461569363277291*10**(-66)` вернет результат идентичный предыдущему, но:

In [15]:
8.461569363277291*10**(-66)

8.461569363277292e-66

Отличие настолько незначительное, что для нас оно становится *абсолютно неважным*. Возможно, поэтому числа типа `float` в Python все чаще называют вещественными, а не числами с плавающей точкой.

In [16]:
def far_to_cels_v1(far):
     return (far - 32) * 5/9
    
def far_to_cels_v2(far):
     return 5/9 * (far - 32)

In [17]:
far_to_cels_v1(456)

235.55555555555554

In [18]:
far_to_cels_v2(456)

235.55555555555557

Давайте посмотрим вот на такой пример:

In [19]:
0.11 + 0.29

0.39999999999999997

Должно получиться ровно $0.4$, а получилось $0.39999999999999997$. 

Конечно, как вы сказали: на такую погрешность можно вообще не обращать внимания, но как минимум, такой результат сложения кажется странным сам по себе. Ну в самом деле, разве это так трудно правильно сложить? Дело в том, что компьютер использует двоичную арифметику, над числами в двоичном представлении, а конечная десятичная дробь, в двоичном представлении может оказаться бесконечной, бесконечный "хвост" которой и отбрасывается при вычислениях, что в свою очередь и приводит к таким "ничтожным" погрешностям.

In [20]:
print("0.1 = {0:.17f}".format(0.1))
print("0.2 = {0:.17f}".format(0.2))
print("0.3 = {0:.17f}".format(0.3))

0.1 = 0.10000000000000001
0.2 = 0.20000000000000001
0.3 = 0.29999999999999999


Могут возникать странности при попытке округления:

In [21]:
round(2.675, 2)

2.67

In [22]:
print("2.675 на самом деле = {0:.17f}".format(2.675))

2.675 на самом деле = 2.67499999999999982


Часто возникает необходимость превратить вещественное число в целое ("округлить"). В питоне есть несколько способов это сделать, но, к сожалению, ни один из них не работает как наше привычное округление и про это всегда нужно помнить.

Большинство этих функций не реализованы в базовом наборе команд питона и для того, чтобы ими пользоваться, нам придется загрузить дополнительную библиотеку math, которая содержит всякие специальные функции для математических вычислений.

In [23]:
import math # команда import загружает модуль под названием math

Самый простой способ округлить число - применить к нему функцию int.

Обратите внимание, что такой метод просто обрубает дробную часть (значения выше 0.5 не округляются в сторону большего числа).

In [24]:
print(int(2.6))
print(int(-2.6))

2
-2


Округление "в пол" из модуля math округляет до ближайшего меньшего целого числа.

In [25]:
print(math.floor(2.6)) # чтобы использовать функцю из дополнительного модуля - 
                        # нужно сначала написать название этого модуля и через точку название функции
print(math.floor(-2.6))

2
-3


Округление "в потолок" работает ровно наоброт - округляет до ближайшего большего числа, независимо от значения дробной части.

In [26]:
print(math.ceil(2.6))
print(math.ceil(-2.6))

3
-2


В самом питоне есть еще функция round(). Вот она работает почти привычно нам, если бы не одно "но"...

In [27]:
print(round(2.2))
print(round(2.7))
print(round(2.5)) # внимание на эту строку
print(round(3.5))

2
3
2
4


Неожиданно? Тут дело в том, что в питоне реализованы не совсем привычные нам правила округления чисел с вещественной частью 0.5 - такое число округляется до ближайшего четного числа: 2 для 2.5 и 4 для 3.5.

### Видео:
    
* https://youtu.be/BotL6wYm5Hc
* https://youtu.be/ZgH3n9drexE
* http://zealcomputing.ru/besedy/003/
* https://vimeo.com/401828406
* https://vimeo.com/401828248
* https://vimeo.com/401828268
* https://vimeo.com/401828284
* https://vimeo.com/401828316
* https://vimeo.com/401828355
* https://vimeo.com/401828377    

#### IEEE-754

![float](https://upload.wikimedia.org/wikipedia/commons/thumb/e/e8/IEEE_754_Single_Floating_Point_Format.svg/618px-IEEE_754_Single_Floating_Point_Format.svg.png)

![double](https://upload.wikimedia.org/wikipedia/commons/thumb/a/a9/IEEE_754_Double_Floating_Point_Format.svg/618px-IEEE_754_Double_Floating_Point_Format.svg.png)

Но, как говорится "Дьявол кроется в мелочах". Очень неприятным последствием таких "ничтожно-маленьких" погрешностей является то, что вы не можете точно проверить истинность равенства:

In [28]:
0.7 + 0.2 - 0.9 == 0

False

Потому что с точки зрения компьютера:

In [29]:
0.7 + 0.2 - 0.9

-1.1102230246251565e-16

Мы привыкли думать о числах в десятичной (по основанию 10) нотации, так что каждая дробь должна быть выражена как сумма степеней из 10:

$1/8 = 1*10^{-1} + 2*10^{-2} + 5*10^{-3}$

В знакомом представлении по основанию 10 мы представляем это в знакомое десятичное выражение: 0,125.
Компьютеры обычно хранят значения в двоичной записи, поэтому каждое число выражается суммой степеней 2:

$1/8 = 0 * 2^{-1} + 0*2^{-2} + 1*2^{-3}$

В представлении по основанию 2 мы можем записать это 0,001 (2), где нижний индекс 2 указывает двоичную запись. Значение 0,125 = 0,001 (2) оказывается одним числом, которое как двоичная, так и десятичная запись могут представлять в конечном количестве цифр.

В привычном представлении чисел по основанию 10 вы, вероятно, знакомы с числами, которые не могут быть выражены конечным числом цифр. Например, деление 1 на 3 дает в стандартной десятичной записи:

$1/3 = 0.333333333...$

... чтобы действительно представить этот коэффициент, количество требуемых цифр бесконечно! 

Точно так же существуют числа, для которых двоичные представления требуется бесконечное количество цифр. Например:

$1/10 = 0.00011001100110011...$ (2)

Точно так же, как десятичное представление требует бесконечного числа цифр для идеального представления $1/3$, двоичное представление требует бесконечного числа цифры представляют $1/10$. Python внутренне усекает эти представления на 52 бита за пределами первого ненулевого бита в большинстве систем.

Эта ошибка округления для значений с плавающей точкой является необходимым злом при работе с числами с плавающей точкой. Лучший способ справиться с этим - всегда иметь в виду, что арифметика с плавающей точкой является приблизительной, и никогда не полагаться на тесты на точное равенство со значениями с плавающей точкой.

А в финансовой, бухгалтерской, ML и гео- средах подобные логические проверки выполняются постоянно.

Вторым неприятным последствием становится то, что погрешности имеют свойство накопления. Расмотрим простой пример:

In [ ]:
s = 0
for i in range(100000000):
    s += 0.1

In [ ]:
s

Мы $100000000$ раз сложили число $0.1$ с самим собой, но вместо $10000000$ мы получили $9999999.98112945$, которое отличается от правильного результата на целых $0.018870549276471138$. В принципе не так уж и сильно отличается. Да и пример "притянут за уши". Но что-то подобное происходит при решении дифференциальных уравнений. Если с помощью таких уравнений строится траектория космического аппарата, то из-за такой мизерной погрешности он, конечно, полетит в сторону нужной планеты, но пролетит мимо. А если вы рассчитываете параметры химической реакции, то на компьютере все может выглядеть более чем безобидно, но в действительности, из-за этой мизерной погрешности вполне может произойти взрыв.

Если вам совсем не обойтись без сравнения чисел с плавающей точкой, то можно сделать так: сравнивать не результат сложения и числа, а разность эти двух чисел с каким-то очень маленьким числом (с таким, размер которого будет точно не критичен для нашего вычисления). Например, порог это числа будет разным для каких-то физических вычислений, где важна высокая точность, и сравнения доходов граждан:

In [ ]:
0.2 + 0.1 - 0.3 < 0.000001

см. [Мануал про float](https://docs.python.org/3/tutorial/floatingpoint.html) и на [русском](https://pythoner.name/documentation/tutorial/floatingpoint)

In [ ]:
# https://docs.python.org/3/library/sys.html#sys.float_info

In [ ]:
import sys
sys.float_info.epsilon

[Что нужно знать про арифметику с плавающей запятой](https://habr.com/ru/post/112953/)

### Десятичные дроби (Decimal)

[Подробнее](https://pyprog.pro/python/st_lib/decimal.html)

Потребность в повышенной точности, возникает очень редко, но возникает неспроста. Именно эту потребность и призваны удовлетворить числа типа `Decimal`. Этот тип не является встроенным, а предоставляется модулем `Decimal` из стандартной библиотеки Python ([детали на официальном сайте](https://docs.python.org/3/library/decimal.html#quick-start-tutorial)):

In [ ]:
from decimal import *    #  импортируем модуль
getcontext().prec = 10    #  устанавливаем точность

In [ ]:
getcontext()

In [ ]:
#  Вычислим частное 13/17
Decimal(13) / Decimal(17)

Причем точность может быть настолько большой, насколько позволяет мощность компьютера. Допустим, мы хотим видеть результат с точностью 
$80$ знаков после запятой (хотя можем увидеть и $1000$), вот они:

In [ ]:
getcontext().prec = 80    #  меняем точность

In [ ]:
Decimal(13) / Decimal(17)

Хотелось бы думать, что такая точность доступна абсолютно для всех математических операций и функций, например таких как всякие синусы, косинусы и прочая экзотика. Но нет, слишком хорошо – тоже не хорошо. К тому же все эти и другие функции могут быть получены с помощью базовых математических операций, которые модулем Decimal прекрасно поддерживаются, например:

In [ ]:
Decimal(3).sqrt()    #  квадратный корень из 3

In [ ]:
Decimal(3)**Decimal(1/7)    #  корень 7-й степени

In [ ]:
Decimal(3).ln()    #  натуральный логарифм

In [ ]:
#from decimal import *    #  импортируем модуль

def far_to_cels_v1(far):
     return (Decimal(far) - Decimal(32)) * Decimal(5/9)
    
def far_to_cels_v2(far):
     return Decimal(5/9) * (Decimal(far) - Decimal(32))

In [ ]:
far_to_cels_v1(456)

In [ ]:
far_to_cels_v2(456)

Интересно, что в `Python 2` модуль `Decimal` был написал на Python, исходники [здесь](https://github.com/python/cpython/blob/2.7/Lib/decimal.py), а в 3-ей версии переписан на Си.

### Обыкновенные дроби (Fraction)

[Подробнее](https://pyprog.pro/python/st_lib/fractions.html)

Рациональные числа, они же - обыкновенные дроби предоставляются модулем `fractions`. Обыкновенная дробь в данном модуле представляется в виде пары двух чисел `numerator` – числитель и `denominator` – знаменатель:

In [ ]:
from fractions import Fraction
a = Fraction(21, 49)
a

Честно говоря без чисел типа `Fraction` можно легко обойтись, но из примера видно, что данный модуль выполнил сокращение числителя и знаменателя автоматически, что довольно любопытно и наводит на вопрос "А где бы мне это могло пригодиться?". 

Самый очевидный ответ – числовые ряды и пределы. Для примера рассмотрим [ряд Лейбница](https://ru.wikipedia.org/wiki/%D0%A0%D1%8F%D0%B4_%D0%9B%D0%B5%D0%B9%D0%B1%D0%BD%D0%B8%D1%86%D0%B0), который сходится к $\pi/4$ (правда медленно... ооочень медленно сходится):

$1 - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} - \frac{1}{11} + \frac{1}{13} - \frac{1}{15} + \frac{1}{17} - \frac{1}{19} + \frac{1}{21} - \cdots = \sum_{n=0}^\infty \, \frac{(-1)^n}{2n+1}$

In [ ]:
for n in range(20):
    print(Fraction((-1)**n, 2*n + 1), end=', ')

Или посмотреть на поведение вот такого предела:

$\pi=\lim \limits_{m\rightarrow \infty }{\frac { (m!)^{4}\,{2}^{4m}}{\left[ (2m )! \right] ^{2}\,m}}$

который тоже можно выразить с помощью чисел типа `fractions`:

In [ ]:
from math import factorial

for m in range(1, 20):
    pi = Fraction(factorial(m)**4*2**(4*m), factorial(2*m)**2*m)
    print(pi, '=', pi.numerator / pi.denominator)

In [ ]:
from fractions import Fraction

def far_to_cels_v1(far):
     return (far - 32) * Fraction(5/9)
    
def far_to_cels_v2(far):
     return Fraction(5/9) * (far - 32)

In [ ]:
far_to_cels_v1(245)

In [ ]:
far_to_cels_v2(245)

In [ ]:
float(Fraction(532925955905508717, 4503599627370496))

### Литература:

[Оригинал © Семён Лукашевский](https://pyprog.pro/python/py/nums/nums.html)